apps/extension/src/storage/base.ts

```diff
@@ -14,75 +15,71 @@ export interface IStorage {
   };
 }
 
-export interface StorageItem<T> {
-  version: string;
-  value: T;
-}
+export type MigrationFn<OldState, NewState> = (prev: OldState) => NewState | Promise<NewState>;
 
-type Version = string;
+export type ExtensionStorageDefaults<T extends { dbVersion: number }> = Required<
+  Omit<T, 'dbVersion'>
+>;
 
-export type MigrationMap<OldState, NewState> = {
-  [K in keyof OldState & keyof NewState]?: (
-    prev: OldState[K],
-    get: <K extends keyof NewState>(key: K) => Promise<NewState[K]>,
-  ) => NewState[K] | Promise<NewState[K]>;
-};
+export interface RequiredMigrations {
+  // Explicitly require key '0' representing...
+  // It is quite difficult writing a generic that covers all migration function kinds.
+  // Therefore, the writer of the migration should ensure it is typesafe when they define it.
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  0: MigrationFn<any, any>;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  [key: number]: MigrationFn<any, any>; // Additional numeric keys
+}
 
-// It is quite difficult writing a generic that covers all migration function kinds.
-// Therefore, the writer of the migration should ensure it is typesafe when they define it.
-// eslint-disable-next-line @typescript-eslint/no-explicit-any
-export type Migrations<T> = Partial<Record<Version, MigrationMap<any, T>>>;
+export interface Version {
+  current: 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10; // and so on, just not zero
+  migrations: RequiredMigrations;
+}
 
-export type VersionSteps = Record<Version, Version>;
+export interface ExtensionStorageProps<T extends { dbVersion: number }> {
+  storage: IStorage;
+  defaults: ExtensionStorageDefaults<T>;
+  version: Version;
+}
 
-export class ExtensionStorage<T> {
-  private migrationLocks: Record<string, Promise<T[keyof T]>> = {};
+export class ExtensionStorage<T extends { dbVersion: number }> {
```

Comment:
can we include a description of the properties for the ExtensionStorage class:
- responsible for managing the storage system – storage versioning and migrations
- the state managed by this class explicitly includes versioning information from `dbVersion`